### PR TITLE
add third party github login for future updates in vscode-leetcode

### PR DESCRIPTION
**Pull request by yihong0618**
Add third party github login for future updates in vscode-leetcode and somehow solve the recaptcha problem.
- This pr [add feature for](https://github.com/jdneo/vscode-leetcode/issues/475)  vscode-leetcode third party auth.
- somehow can solve the recaptcha  problem for [leetcode-vscode can't login](https://github.com/jdneo/vscode-leetcode/issues/478)

Users only need to connect github(or already connected) can use **leetcode user -g**  can login leetcode in leetcode-cli and future for vscode-leetcode(other third parties also can add except google... I will add it if necessary)

how to connect github

1. find profile page (https://leetcode.com/profile/account/)
2. link your github id like below
3. Or user use github to login leetcode at the first will also work
4. use github account can login to leetcode
image

### CODE REVIEW

**Comment by jdneo** [comment]
Will take a look at the PR before the end of this week. Thank you @yihong0618

File: lib/commands/user.js [hunk 1]
```diff
@@ -66,6 +73,21 @@ cmd.handler = function(argv) {
...
+    prompt.message = '';
+    prompt.start();
+    prompt.get([
+      {name: 'login', required: true},
+      {name: 'pass', required: true, hidden: true}
+    ], function(e, user) {
+      if (e) return log.fail(e)
+      core.githubLogin(user, function(e, user) {
+      if (e) return log.fail(e);
+      log.info('Successfully github login as', chalk.yellow(user.name));
```
[the remark this answers is not in the document]
Comment (yihong0618) [reply lib/commands/user.js]: Yes.

**Comment by yihong0618** [comment]
And I tonight I will add facebook and linkedin login maybe, Google will not work.
I think in vscode-leetcode just one third party login will be nice, user can just select their way to login.
And just to change singin method in vscode-leetcode will be easy.

**Comment by yihong0618** [comment]
Is this one pr OK? Or I start another pr for the whole changes?

**Comment by jdneo** [comment]
Hmm, actually I failed to login using my GitHub account. Does your GitHub account enable the two-factor auth?

**Comment by yihong0618** [comment]
> Hmm, actually I failed to login using my GitHub account. Does your GitHub account enable the two-factor auth?
Sorry,
Mine is fine. I will test another account.
1. No my account not enable  two-factor auth
2. Did you use your Github accout not your leetcode accout
3. Did you link your Github accout to leetcode in leetcode.
4. Can you try to linkedin accout either, because I only have one, sorry for that.

**Comment by jdneo** [comment]
Sure I'll try with LinkedIn later tonight...

**Comment by yihong0618** [comment]
image
I sign up a new accout that use github accout at the first time.
If its a new accout you must verity your email in leetcode and email accout, maybe you miss that?

**Comment by jdneo** [comment]
It should be verified before. The account already connected.

image

Could you try enable the 2-factor auth for your GitHub account? You can find it in the Settings > Security

**Comment by yihong0618** [comment]
> It should be verified before. The account already connected.
> 
> image
> 
> Could you try enable the 2-factor auth for your GitHub account? You can find it in the Settings > Security

Trying.....
But I found no +86(China) for sms

**Comment by yihong0618** [comment]
> It should be verified before. The account already connected.
> 
> image
> 
> Could you try enable the 2-factor auth for your GitHub account? You can find it in the Settings > Security

And I think you can disconnect  and connect another Github account for test.
Before that you should logout your github account in chrome because leetcode connect will read your session

**Comment by jdneo** [comment]
Tried another GitHub account without 2-factor auth, and login successfully.

So I think we can first hint to user that 2-factor auth is not supported yet. And spend some time to see if it is possible to enable it.

File: lib/plugins/leetcode.js [hunk 2]
```diff
@@ -538,19 +538,118 @@ plugin.login = function(user, cb) {
...
+      },
+    };
+    _request(options, function(e, resp, body) {
+      if (resp.statusCode !== 200) {
+        return cb('GitHub login failed');
+      }
+      _request.get({url: leetcodeUrl}, function(e, resp, body) {
+        const redirectUri = resp.request.uri.href;
+        if (redirectUri !== 'https://leetcode.com/') {
+          return cb('GitHub login failed or GitHub did not link to leetcode');
```
Comment (jdneo) [inline lib/plugins/leetcode.js]: leetcode -> LeetCode

File: lib/plugins/leetcode.js [hunk 3]
```diff
@@ -538,19 +538,118 @@ plugin.login = function(user, cb) {
...
+  const leetcodeUrl = config.sys.urls.linkined_login;
+  const _request = request.defaults({
+    jar:     true,
+    headers: {
+      'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/73.0.3683.86 Safari/537.36'
+    }
+  });
+  _request('https://www.linkedin.com', function(e, resp, body) {
+    if ( resp.statusCode !== 200) {
+      return cb('Get linkedin session failed');
```
Comment (jdneo) [inline lib/plugins/leetcode.js]: linkedin -> LinkedIn

File: lib/plugins/leetcode.js [hunk 4]
```diff
@@ -538,19 +538,118 @@ plugin.login = function(user, cb) {
...
+    }
+    const authenticityToken = body.match(/input name="loginCsrfParam" value="(.*)" /);
+    if (authenticityToken === null) {
+      return cb('Get Linkedin token failed');
```
Comment (jdneo) [inline lib/plugins/leetcode.js]: LinkedIn

File: lib/plugins/leetcode.js [hunk 5]
```diff
@@ -538,19 +538,118 @@ plugin.login = function(user, cb) {
...
+      },
+    };
+    _request(options, function(e, resp, body) {
+      if (resp.statusCode !== 200) {
+        return cb('Linkedin login failed');
+      }
+      _request.get({url: leetcodeUrl}, function(e, resp, body) {
+        const redirectUri = resp.request.uri.href;
+        if (redirectUri !== 'https://leetcode.com/') {
+          return cb('Linkedin login failed or Linkedin did not link to leetcode');
```
Comment (jdneo) [inline lib/plugins/leetcode.js]: LinkedIn

LeetCode

**Comment by yihong0618** [comment]
> Tried another GitHub account without 2-factor auth, and login successfully.
> 
> So I think we can first hint to user that 2-factor auth is not supported yet. And spend some time to see if it is possible to enable it.

OK, seems a little hard but I will try. 
And facebook way I think it is not worth because I use **m.facebook.com** a hack way to login.
Although successful but cause the facebook security alert, so I deleted the method.

File: lib/plugins/leetcode.js (hunk 5 above)
Comment (yihong0618) [reply lib/plugins/leetcode.js]: Sorry for that, I will correct them~
Thank you very much for your help ~

**Comment by jdneo** [comment]
Yes, no need to support facebook at this time. GitHub and LinkedIn are good enough!

**Comment by jdneo** [comment]
Thank you @yihong0618.

Would you like me to publish a new release for this change?

**Comment by yihong0618** [comment]
Yes.
And I found a way to solve the two-factor code. But it will use another package.
And I am not sure if it could be work in vscode-leetcode.
So I will have another pr, can you help me with that to check if the two-factor can be used in vscode-leetcode.
Thank you very much. The new pr will send in 10-20 minutes.
Thanks again.

**Comment by jdneo** [comment]
@yihong0618 Sure! Will the PR you mentioned be raised in this repo?

**Comment by yihong0618** [comment]
@jdneo 
Yes, see this https://github.com/leetcode-tools/leetcode-cli/pull/35
Thank you very much!

**Pull request merged**